tcptracer-bpf.c

```diff
@@ -0,0 +1,961 @@
+	updated_status.offset_saddr = status->offset_saddr;
+	updated_status.offset_daddr = status->offset_daddr;
+	updated_status.offset_sport = status->offset_sport;
+	updated_status.offset_dport = status->offset_dport;
+	updated_status.offset_netns = status->offset_netns;
+	updated_status.offset_ino = status->offset_ino;
+	updated_status.offset_family = status->offset_family;
+	updated_status.offset_daddr_ipv6 = status->offset_daddr_ipv6;
+	updated_status.err = 0;
+	updated_status.saddr = status->saddr;
+	updated_status.daddr = status->daddr;
+	updated_status.sport = status->sport;
+	updated_status.dport = status->dport;
+	updated_status.netns = status->netns;
+	updated_status.family = status->family;
+
+	int i;
+	for (i = 0; i < 4; i++) {
+		updated_status.daddr_ipv6[i] = status->daddr_ipv6[i];
+	}
+
+	u32 possible_daddr_ipv6[4] = { };
+	switch (status->what) {
+		case GUESS_DADDR_IPV6:
+			bpf_probe_read(&possible_daddr_ipv6, sizeof(possible_daddr_ipv6), ((char *)skp) + status->offset_daddr_ipv6);
+
+			int i;
+			for (i = 0; i < 4; i++) {
+				updated_status.daddr_ipv6[i] = possible_daddr_ipv6[i];
+			}
+			break;
+		default:
+			// not for us
+			return 0;
+	}
+
+	bpf_map_update_elem(&tcptracer_status, &zero, &updated_status, BPF_ANY);
+
+	return 0;
+}
+
+static inline bool check_family(struct sock *sk, u16 expected_family) {
+	struct tcptracer_status_t *status;
+	u64 zero = 0;
+	u16 family;
+	family = 0;
+
+	status = bpf_map_lookup_elem(&tcptracer_status, &zero);
+	if (status == NULL || status->status != TCPTRACER_STATUS_READY) {
+		return 1;
+	}
+
+	bpf_probe_read(&family, sizeof(u16), ((char *)sk) + status->offset_family);
+
+	return family == expected_family;
+}
+
+SEC("kprobe/tcp_v4_connect")
+int kprobe__tcp_v4_connect(struct pt_regs *ctx)
+{
+	struct sock *sk;
+	u64 pid = bpf_get_current_pid_tgid();
+
+	sk = (struct sock *) PT_REGS_PARM1(ctx);
+
+	bpf_map_update_elem(&connectsock_ipv4, &pid, &sk, BPF_ANY);
+
+	return 0;
+}
+
+SEC("kretprobe/tcp_v4_connect")
+int kretprobe__tcp_v4_connect(struct pt_regs *ctx)
+{
+	int ret = PT_REGS_RC(ctx);
+	u64 pid = bpf_get_current_pid_tgid();
+	struct sock **skpp;
+	u64 zero = 0;
+	struct tcptracer_status_t *status;
+
+	skpp = bpf_map_lookup_elem(&connectsock_ipv4, &pid);
+	if (skpp == 0) {
+		return 0;	// missed entry
+	}
+
+	struct sock *skp = *skpp;
+
+	bpf_map_delete_elem(&connectsock_ipv4, &pid);
+
+	if (ret != 0) {
+		// failed to send SYNC packet, may not have populated
+		// socket __sk_common.{skc_rcv_saddr, ...}
+		return 0;
+	}
+
+	status = bpf_map_lookup_elem(&tcptracer_status, &zero);
+	if (status == NULL || status->status == TCPTRACER_STATUS_UNINITIALIZED) {
+		return 0;
+	}
+
+	if (!are_offsets_ready_v4(status, skp, pid)) {
+		return 0;
+	}
+
+	// pull in details
+	struct ns_common *ns;
+	u32 saddr, daddr, net_ns_inum;
+	u16 sport, dport;
+
+	sport = 0;
+	saddr = 0;
+	daddr = 0;
+	dport = 0;
+	bpf_probe_read(&sport, sizeof(sport), ((char *)skp) + status->offset_sport);
```

Comment:
In case of an error, `bpf_probe_read()` [sets the output to zeros](http://lxr.free-electrons.com/source/kernel/trace/bpf_trace.c#L64) and returns the kind of error, which [is always `-EFAULT`](http://lxr.free-electrons.com/source/mm/maccess.c#L38?v=4.9).

We could print those errors with `bpf_trace_printk()` but they will appear on `/sys/kernel/debug/tracing/trace_pipe` which is not very useful and the purpose of that function is debugging.

I think it's ok to not check errors and it seems a [common pattern](https://github.com/iovisor/bcc/search?utf8=%E2%9C%93&q=bpf_probe_read)